ruby/ql/lib/codeql/ruby/regexp/RegExpInjectionCustomizations.qll

```diff
@@ -0,0 +1,78 @@
+/**
+ * Provides default sources, sinks and sanitizers for reasoning about regexp
+ * injection vulnerabilities, as well as extension points for adding your own.
+ */
+
+private import ruby
+private import codeql.ruby.DataFlow
+private import codeql.ruby.Concepts
+private import codeql.ruby.Frameworks
+private import codeql.ruby.dataflow.RemoteFlowSources
+private import codeql.ruby.dataflow.BarrierGuards
+private import codeql.ruby.ApiGraphs
+
+/**
+ * Provides default sources, sinks and sanitizers for detecting
+ * regexp injection vulnerabilities, as well as extension points for
+ * adding your own.
+ */
+module RegExpInjection {
+  /**
+   * A data flow source for regexp injection vulnerabilities.
+   */
+  abstract class Source extends DataFlow::Node { }
+
+  /**
+   * A data flow sink for regexp injection vulnerabilities.
+   */
+  abstract class Sink extends DataFlow::Node { }
+
+  /**
+   * A sanitizer guard for regexp injection vulnerabilities.
+   */
+  abstract class SanitizerGuard extends DataFlow::BarrierGuard { }
+
+  /**
+   * A data flow sanitized for regexp injection vulnerabilities.
+   */
+  abstract class Sanitizer extends DataFlow::Node { }
+
+  /**
+   * A source of remote user input, considered as a flow source.
+   */
+  class RemoteFlowSourceAsSource extends Source, RemoteFlowSource { }
+
+  /** A regexp literal, considered as a flow sink. */
+  class RegExpLiteralAsSink extends Sink {
+    RegExpLiteralAsSink() { this.asExpr().getExpr() instanceof RegExpLiteral }
+  }
+
+  /**
+   * The first argument of a call to `Regexp.new`, considered as a flow sink.
+   */
+  class ConstructedRegExpAsSink extends Sink {
+    ConstructedRegExpAsSink() {
+      this =
+        API::getTopLevelMember("Regexp").getAnInstantiation().(DataFlow::CallNode).getArgument(0)
```

Comment:
There are a few other ways I've found for creating regexes:
- [`Regexp.compile`](https://ruby-doc.org/core-3.0.2/Regexp.html#method-c-compile), which is an alias for `Regexp.new`
- [`Regexp.union`](https://ruby-doc.org/core-3.0.2/Regexp.html#method-c-union), which will create a regex matching a union of its arguments